# Data Validation
The objective of this stage is to transform the raw datasets into a standardized, comparable format. Before performing the actual reconciliation (comparison), we must ensure that both files are technically consistent and free from structural "noise."

This step include: 
- missing data verification - identifying gaps (nulls) in critical columns like `UnitPrice` or `CustomerID`. We assess their impact and decide whether to impute, flag, or exclude them from the financial comparison.
- data type standardization - ensuring that key columns used for matching (e.g., `CustomerID`, `InvoiceNo`) have consistent data types in both datasets (converting strings to numeric or vice versa) to avoid join errors.
- formatting standardization - cleaning technical "blockers" (hidden spaces, unifying capitalziation, fixing corrupted data formats)
- duplicate detection and isolation - identifying redundant records that could distort totals. Technical duplicates will be moved to separate dataframes for audit purpose, while the main datasets are deduplicated to establish a "Single Source of Truth".

Purpose: To reach a state where both datasets are technically "aligned," allowing for an accurate row-by-row comparison in the Comparison phase.

**Step 1** Copy of both datasets and load.

In [5]:
import pandas as pd
import os

#folder for copied files
os.makedirs('validation', exist_ok=True)

def load_data(file_path):
    try:
        df = pd.read_csv(file_path, encoding='latin1')
        print(f"File {file_path} loaded correctly using latin1 encoding. Loaded {len(df)} rows.")
        return df
    except Exception:
        df = pd.read_csv(file_path, encoding='cp1252')
        print(f"File {file_path} loaded correctly using cp1252 encoding. Loaded {len(df)} rows.")
        return df

# oryginal files
df_oms_org = load_data('oms_orders_placed.csv')
df_wms_org = load_data('wms_shipped_units.csv')

# saving and loadin copy of files 
df_oms_org.to_csv('validation/oms_orders_placed_copy.csv', index=False, encoding='latin1')
df_wms_org.to_csv('validation/wms_shipped_units_copy.csv', index=False, encoding='latin1')

df_oms = pd.read_csv('validation/oms_orders_placed_copy.csv', encoding='latin1')
df_wms = pd.read_csv('validation/wms_shipped_units_copy.csv', encoding='latin1')

File oms_orders_placed.csv loaded correctly using latin1 encoding. Loaded 541909 rows.
File wms_shipped_units.csv loaded correctly using latin1 encoding. Loaded 542409 rows.


**Step 2** Flags creation

In [6]:
# Initializing the status and flags column
for df in [df_oms, df_wms]:
    if 'validation_status' not in df.columns:
        df['validation_status'] = 'CLEAN'
    df['validation_flags'] = [[] for _ in range(len(df))]

def add_flag(df, mask, flag_name):
    df.loc[mask, 'validation_flags'] = df.loc[mask, 'validation_flags'].apply(lambda flags: flags + [flag_name])
    return df
def assign_status(flags):
    if any(f in flags for f in ['Invalid_Date', 'Missing_UnitPrice']):
        return 'ERROR'
    elif 'Duplicate_Record' in flags:
        return 'DUPLICATE'
    elif 'Missing_Description' in flags:
        return 'WARNING'
    else:
        return 'CLEAN'

**Step 2** Missing data verification

In [7]:
def check_missing_data(df, name):
    print(f"Missing Data Analysis: {name}")
    missing = df.isnull().sum()
    missing_pct = (df.isnull().sum()/len(df)) * 100
    
    missing_df = pd.DataFrame({
        'Missing values': missing,
        'Percentage (%)': missing_pct
    })
    
    print(missing_df[missing_df['Missing values'] > 0])
    return missing_df

#running verification for both files
oms_missing = check_missing_data(df_oms, "OMS")
wms_missing = check_missing_data(df_wms, "WMS")


Missing Data Analysis: OMS
             Missing values  Percentage (%)
Description            1454        0.268311
CustomerID           135080       24.926694
Missing Data Analysis: WMS
             Missing values  Percentage (%)
Description            1454        0.268063
UnitPrice              1000        0.184363
CustomerID           134660       24.826284


Observation: 
- both datasets have same amount of missing values in Description
- both datasets are missing CustomerID, but in 'original' OMS the value is higher
- WMS dataset is missing 1000 UnitPrice

**Step 3** Data type verification adn standarization

In [8]:
df_oms.info()
df_wms.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   InvoiceNo          541909 non-null  object 
 1   StockCode          541909 non-null  object 
 2   Description        540455 non-null  object 
 3   Quantity           541909 non-null  int64  
 4   InvoiceDate        541909 non-null  object 
 5   UnitPrice          541909 non-null  float64
 6   CustomerID         406829 non-null  float64
 7   Country            541909 non-null  object 
 8   validation_status  541909 non-null  object 
 9   validation_flags   541909 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 41.3+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 542409 entries, 0 to 542408
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   InvoiceNo          54240

In [9]:
# date validation
df_oms['InvoiceDate'] = pd.to_datetime(df_oms['InvoiceDate'], errors = 'coerce')
df_wms['InvoiceDate'] = pd.to_datetime(df_wms['InvoiceDate'], errors = 'coerce')

print(f"Incorrect dates in OMS: {df_oms['InvoiceDate'].isna().sum()}")
print(f"Incorrect dates in WMS: {df_wms['InvoiceDate'].isna().sum()}")

date_error_mask_oms = df_oms['InvoiceDate'].isna()
date_error_mask_wms = df_wms['InvoiceDate'].isna()

df_oms = add_flag(df_oms, date_error_mask_oms, 'Invalid_Date')
df_wms = add_flag(df_wms, date_error_mask_wms, 'Invalid_Date')

# audit files for incorrect dates
df_oms[date_error_mask_oms].to_csv('validation/oms_date_errors_audit.csv', index=False)
df_wms[date_error_mask_wms].to_csv('validation/wms_date_errors_audit.csv', index=False)

C:\Users\marty\AppData\Local\Temp\ipykernel_26576\1211945731.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_wms['InvoiceDate'] = pd.to_datetime(df_wms['InvoiceDate'], errors = 'coerce')


Incorrect dates in OMS: 0
Incorrect dates in WMS: 11


Observation:
- 11 incorrect dates in WMS identified. I'll move this 11 records to another file.

**Step 4** Data type and formatting standarization

In [10]:
def clean_and_standardize(df):
    if 'validation_status' not in df.columns:
        df['validation_status'] = 'CLEAN'
    if 'validation_flags' not in df.columns:
        df['validation_flags'] = [[] for _ in range(len(df))]
        
    # use pd.to_numeric first to handle strings like '13047.0' then convert to Int to remove decimals, then to String for matching.
    df['CustomerID'] = pd.to_numeric(df['CustomerID'], errors='coerce') \
                        .fillna(0) \
                        .astype(int) \
                        .astype(str) \
                        .replace('0', 'Unknown')
    
    # Removing hidden spaces (strip) and unifying capitalization (upper)
    text_columns = ['InvoiceNo', 'StockCode', 'Description', 'Country']
    for col in text_columns:
        df[col] = df[col].astype(str).str.strip().str.upper().fillna('')
    
    # Quantity
    df['Quantity'] = pd.to_numeric(df['Quantity'], errors='coerce').fillna(0).astype(int)
    
    # UnitPrice - missing values flagges as WARNING
    missing_price_mask = df['UnitPrice'].isna()
    df = add_flag(df, missing_price_mask, 'Missing_UnitPrice')
    df['UnitPrice'] = pd.to_numeric(df['UnitPrice'], errors='coerce').fillna(0)
    
    # Description -  missing values flagges as WARNING
    missing_desc_mask = df['Description'].isna()
    df = add_flag(df, missing_desc_mask, 'Missing_Description')
    
    return df

df_oms = clean_and_standardize(df_oms)
df_wms = clean_and_standardize(df_wms)


**Step 5** Duplicates detection

In [ ]:
# choosing columns which define record (without flag columns)
cols_to_check_oms = [c for c in df_oms.columns if c not in ['validation_flags', 'validation_status']]
cols_to_check_wms = [c for c in df_wms.columns if c not in ['validation_flags', 'validation_status']]

# duplicates masks
oms_dup_mask = df_oms.duplicated(subset=cols_to_check_oms, keep=False)
wms_dup_mask = df_wms.duplicated(subset=cols_to_check_wms, keep=False)

# adding flags
df_oms = add_flag(df_oms, oms_dup_mask, 'Duplicate_Record')
df_wms = add_flag(df_wms, wms_dup_mask, 'Duplicate_Record')

# audit files for duplicates
df_oms[oms_dup_mask].to_csv('validation/oms_duplicates_audit.csv', index=False)
df_wms[wms_dup_mask].to_csv('validation/wms_duplicates_audit.csv', index=False)

# assigning main status according to priority
df_oms['validation_status'] = df_oms['validation_flags'].apply(assign_status)
df_wms['validation_status'] = df_wms['validation_flags'].apply(assign_status)


# final files
df_oms.to_csv('validation/oms_final_clean.csv', index=False)
df_wms.to_csv('validation/wms_final_clean.csv', index=False)


print("\nVALIDATION SUMMARY REPORT")

for name, df in [('OMS', df_oms), ('WMS', df_wms)]:
    print(f"\n{name} Dataset Status:")
    print(f"  CLEAN records: {(df['validation_status']=='CLEAN').sum()}")
    print(f"  ERROR records: {(df['validation_status']=='ERROR').sum()}")
    print(f"  DUPLICATE records: {(df['validation_status']=='DUPLICATE').sum()}")
    print(f"  WARNING records: {(df['validation_status']=='WARNING').sum()}")
    print(f"  TOTAL records: {len(df)}")
    print(f"  Unique flags: {df['validation_flags'].explode().unique()}")


VALIDATION SUMMARY REPORT

OMS Dataset Status:
  CLEAN records: 531762
  ERROR records: 0
  DUPLICATE records: 10147
  WARNING records: 0
  TOTAL records: 541909
  Unique flags: [nan 'Duplicate_Record']

WMS Dataset Status:
  CLEAN records: 530316
  ERROR records: 1011
  DUPLICATE records: 11082
  WARNING records: 0
  TOTAL records: 542409
  Unique flags: ['Invalid_Date' nan 'Missing_UnitPrice' 'Duplicate_Record']


: 

**Conclusions**

Missing data verification:
- OMS: 1,454 missing descriptions, 135,080 missing CustomerIDs
- WMS: 1,454 missing descriptions, 134,660 missing CustomerIDs, 1,000 missing UnitPrices
- Missing values flagged as WARNING for detailed tracking during reconciliation

Data type and format standardization:
- Converted InvoiceDate to datetime format; identified 11 corrupted date entries in WMS (ERR_DATE_2024)
- Standardized CustomerID: converted from mixed float/string to consistent string format, removed hidden spaces
- Unified text fields (InvoiceNo, StockCode, Description, Country) to uppercase and stripped whitespace
- Ensured Quantity and UnitPrice are numeric types

Validation Flagging System:
- All records retained with validation flags (CLEAN, ERROR, DUPLICATE, WARNING)
- validation_flags column contains list of all issues for each record
- validation_status shows highest priority issue
- Enables detailed analysis in Comparison phase

Technical Data Flagging:
- Flagged 11 date errors in WMS (Invalid_Date flag)
- Identified and flagged duplicate records:
  - OMS: 5,268 duplicate rows flagged
  - WMS: 5,702 duplicate rows flagged (434 more than OMS)
- All records retained for comprehensive comparison

Final Dataset Status:
- OMS final: 541,909 records (all retained with flags)
- WMS final: 542,409 records (all retained with flags)
- Discrepancy: 500 records (to be investigated in Comparison phase)

Audit Files Created:
- wms_date_errors_audit.csv - 11 records with corrupted dates
- oms_duplicates_audit.csv - 5,268 duplicate records
- wms_duplicates_audit.csv - 5,702 duplicate records